scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+        SELECT rid FROM ridsOfAllFids EXCEPT DISTINCT SELECT rid FROM (SELECT MAX(rid) AS rid, fid FROM ridsOfLatestFids GROUP BY fid)
+        """
+
+        return self._hit_bigquery(query)
+
+    def get_null_filerefs(self, target_table: str) -> Set[str]:
+        """
+        Determines what rows have null values in the file_id column. We want to soft delete those.
+        :param target_table: The particular table to operate on.
+        :return: A set of row ids to soft delete.
+        """
+        query = f"""
+        SELECT datarepo_row_id
+        FROM `{self.project}.datarepo_{self.dataset}.{target_table}` WHERE file_id IS NULL
+        """
+
+        return self._hit_bigquery(query)
+
+    def _hit_bigquery(self, query):
+        """
+        Helper function to consistently interact with biqquery while reusing the same client.
+        :param query: The SQL query to run.
+        :return: A set of whatever the query is asking for (assumes that we're only asking for a single column).
+        """
+        query_job = self.bigquery_client.query(query)
+        return {row[0] for row in query_job}
+
+    # local csv interactions
+    def create_csv(self, row_ids: Set[str], target_table: str) -> str:
+        """
+        Create a csv locally with one column filled with row ids to soft delete.
+        :param row_ids: A set of row ids to soft delete.
+        :param target_table: The table that the row ids belong to.
+        :return: The filename of the created csv.
+        """
+        filename = self.filename_template.format(table=target_table)
+        with open(filename, mode="w") as soft_delete_file:
+            sd_writer = csv.writer(soft_delete_file, delimiter=",", quotechar='"', quoting=csv.QUOTE_MINIMAL)
+
+            for rid in row_ids:
+                sd_writer.writerow([rid])
+
+        return filename
+
+    @staticmethod
+    def delete_csv(filename: str):
+        """
+        Remove the csv after it has been uploaded so that there isn't a deluge of csv files in the local directory.
+        :param filename: The filename of the file to delete.
+        :return:
+        """
+        os.remove(filename)
+
+    # gcs (cloud storage) interactions
+    def put_csv_in_bucket(self, filename: str) -> str:
+        """
+        Puts a local file into a GCS bucket accessible by Jade.
+        :param filename: The filename of the file to upload.
+        :return: The gs-path of the uploaded file.
+        """
+        storage_client = storage.Client(project=self.bucket_project, credentials=self.gcp_creds)
+        bucket = storage_client.bucket(self.bucket)
+        blob = bucket.blob(filename)
+        blob.upload_from_filename(filename)
+
+        filepath = f"gs://{self.bucket}/{filename}"
+        print(f"Put a soft-delete file here: {filepath}")
+
+        return filepath
+
+    # jade interactions
+    @staticmethod
+    @lru_cache()
```

Comment:
❓ Is this caching needed, i.e., what are the performance issues this is addressing?

Reply:
Probably not needed, just figured that since the API structure doesn't change very often we don't want to create this map for every function invocation

Reply:
I suggest removing it then; caching unless it's absolutely necessary is a sure route to hard to diagnose bugs.